src/components/CodeShape/index.spec.tsx

```diff
@@ -0,0 +1,232 @@
+import * as React from 'react';
+import { shallow } from 'enzyme';
+
+import styles from './styles.module.scss';
+
+import CodeShape, {
+  PublicProps as CodeShapeProps,
+  Token,
+  generateLineShapes,
+} from '.';
+
+describe(__filename, () => {
+  describe('generateLineShapes', () => {
+    it('collects token counts', () => {
+      const use = 'use';
+      const space = ' ';
+      const strict = 'strict;';
+      const line = `${use}${space}${strict}`;
+
+      expect(generateLineShapes([line])).toEqual([
+        {
+          line: 1,
+          tokens: [
+            expect.objectContaining({ token: Token.code, count: use.length }),
+            expect.objectContaining({
+              token: Token.whitespace,
+              count: space.length,
+            }),
+            expect.objectContaining({
+              token: Token.code,
+              count: strict.length,
+            }),
+            // This is the trailing whitespace
+            expect.any(Object),
+          ],
+        },
+      ]);
+    });
+
+    it('collects trailing whitespace', () => {
+      const maxLineLength = 80;
+      const line = 'use strict;';
+
+      const allShapes = generateLineShapes([line], { maxLineLength });
+      expect(allShapes[0].tokens.pop()).toEqual(
+        expect.objectContaining({
+          token: Token.whitespace,
+          // This is a count of the trailing whitespace in relation to the
+          // window size (i.e. maxLineLength)
+          count: maxLineLength - line.length,
+        }),
+      );
+    });
+
+    it('collects leading whitespace', () => {
+      const leadingSpace = '    ';
+      const line = `${leadingSpace}use strict;`;
+
+      const allShapes = generateLineShapes([line]);
+      expect(allShapes[0].tokens[0]).toEqual(
+        expect.objectContaining({
+          token: Token.whitespace,
+          count: leadingSpace.length,
+        }),
+      );
+    });
+
+    it('calculates width percentages', () => {
+      const line = 'AAAA  BB';
+
+      expect(
+        generateLineShapes([line], { maxLineLength: line.length }),
+      ).toEqual([
+        {
+          line: 1,
+          tokens: [
+            // Check 'AAAA', the first 50%
+            expect.objectContaining({ token: Token.code, percentOfWidth: 50 }),
+            // Check ' ', the next 25%
+            expect.objectContaining({
+              token: Token.whitespace,
+              percentOfWidth: 25,
+            }),
+            // Check 'BB', the final 25%
+            expect.objectContaining({ token: Token.code, percentOfWidth: 25 }),
+          ],
+        },
+      ]);
+    });
+
+    it('respects maxLineLength when counting tokens', () => {
+      const line = 'AA  BBB ; the rest will be ignored';
```

Comment:
Oh, so the code is not scaled down according to `maxLineLength`? It is truncated instead, is that correct?

Reply:
Yes, it's truncated. I think this is a good idea for minified files that have really long lines.

Reply:
Sure, I think it's simpler than scaling down the code.